### PR TITLE
Escape subnormal values

### PR DESCRIPTION
**Pull request by shiltian**
Currently we don't escape subnormal values when generating image data. In sampler read tests, we use `!=` to check the two values even when it is floating-point data, which requires the two values are bitwise equal. However, a sampler might flush subnormal values, causing the test case to fail.

In this patch, when generating random image data, we escape subnormal values.

### CODE REVIEW

**Comment by b-sumner** [comment]
Section 8.3.3 of the C spec specifically allows sampling to flush denorms, so this change seems to be long overdue.

**Comment by lakshmih** [comment]
Reviewing

**Review by bashbaug — APPROVED** [review]
I think I'm OK with this, but noting for completeness: this change removes all fp32 and fp16 denorm values from the input data.  If we want to have _some_ coverage for denorm inputs still, we will want to continue to include denorm inputs, and instead we will want to handle cases where denorms are preserved and when denorms are flushed to zero.

**Comment by bashbaug** [comment]
A bit of spec archaeology, for the Khronos folks:
* A fix similar to this one was proposed in Bugzilla 16324.
* The spec changes were discussed in GitLab issues 30 and 113.

**Review by bashbaug — APPROVED** [review]
Re-approving, looks like only whitespace / formatting changed.

**Comment by shiltian** [comment]
I don't have privileges to merge the PR. Can someone do it for me? Thanks!

**Comment by bashbaug** [comment]
It'd be nice to get one more approval before merging - @silverclaw or @lakshmih maybe?

If not, we'll aim to merge in our next teleconference, on Tuesday (June 4th).

**Review by silverclaw — APPROVED** [review]
LGTM

**Review by lakshmih — APPROVED** [review]

**Comment by bashbaug** [comment]
Thanks!  Merging.

**Pull request merged**
